hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/ec/reconstruction/ECReconstructionCoordinator.java

```diff
@@ -0,0 +1,417 @@
+              repConfig, null, indexDnPair.getKey());
+    }
+
+    // 2. Reconstruct and transfer to targets
+    for (BlockLocationInfo blockLocationInfo : blockLocationInfoMap.values()) {
+      reconstructECBlockGroup(blockLocationInfo, repConfig, targetNodeMap);
+    }
+
+    // 3. Close containers
+    iterator = targetIndexDns.iterator();
+    while (iterator.hasNext()) {
+      DatanodeDetails dn = iterator.next().getValue();
+      this.containerOperationClient
+          .closeContainer(containerID, dn, repConfig, null);
+    }
+
+  }
+
+  void reconstructECBlockGroup(BlockLocationInfo blockLocationInfo,
+      ECReplicationConfig repConfig,
+      SortedMap<Integer, DatanodeDetails> targetMap)
+      throws IOException {
+    long safeBlockGroupLength = blockLocationInfo.getLength();
+    List<Integer> missingContainerIndexes =
+        targetMap.keySet().stream().collect(Collectors.toList());
+
+    // calculate the real missing block indexes
+    int dataLocs = ECBlockInputStreamProxy
+        .expectedDataLocations(repConfig, safeBlockGroupLength);
+    List<Integer> toReconstructIndexes = new ArrayList<>();
+    for (int i = 0; i < missingContainerIndexes.size(); i++) {
+      Integer index = missingContainerIndexes.get(i);
+      if (index <= dataLocs || index > repConfig.getData()) {
+        toReconstructIndexes.add(index);
+      }
+      // else padded indexes.
+    }
+
+    // Looks like we don't need to reconstruct any missing blocks in this block
+    // group. The reason for this should be block group had only padding blocks
+    // in the missing locations.
+    if (toReconstructIndexes.size() == 0) {
+      if (LOG.isDebugEnabled()) {
+        LOG.debug("Skipping the reconstruction for the block: "
+            + blockLocationInfo.getBlockID() + ". In the missing locations: "
+            + missingContainerIndexes
+            + ", this block group has only padded blocks.");
+      }
+      return;
+    }
+
+    try (ECBlockReconstructedStripeInputStream sis
+        = new ECBlockReconstructedStripeInputStream(
+        repConfig, blockLocationInfo, true,
+        this.containerOperationClient.getXceiverClientManager(), null,
+        this.blockInputStreamFactory, byteBufferPool,
+        this.ecReconstructExecutor)) {
+
+      ECBlockOutputStream[] targetBlockStreams =
+          new ECBlockOutputStream[toReconstructIndexes.size()];
+      ByteBuffer[] bufs = new ByteBuffer[toReconstructIndexes.size()];
+      for (int i = 0; i < toReconstructIndexes.size(); i++) {
+        OzoneClientConfig configuration = new OzoneClientConfig();
+        // TODO: Let's avoid unnecessary bufferPool creation for
+        BufferPool bufferPool =
+            new BufferPool(configuration.getStreamBufferSize(),
+                (int) (configuration.getStreamBufferMaxSize() / configuration
+                    .getStreamBufferSize()),
+                ByteStringConversion.createByteBufferConversion(false));
```

Comment:
Nit: not sure what the TODO is about, but at least this can be moved outside the loop.

Reply:
Sorry, it was partially written todo, I updated it.